aptos-move/aptos-vm/src/verifier/transaction_arg_validation.rs

```diff
@@ -138,98 +166,80 @@ pub(crate) fn is_valid_txn_arg<S: MoveResolverExt>(
     }
 }
 
-// Validate arguments. Walk through the arguments and according to the signature
-// validate arguments that require so.
-// TODO: This needs a more solid story and a tighter integration with the VM.
-// Validation at the moment is only for Strings and Vector of them, so we
-// manually walk the serialized args until we find a string.
-// This is obviously brittle and something to change at some point soon.
-pub(crate) fn validate_args<S: MoveResolverExt>(
-    session: &SessionExt<S>,
-    idxs: &[usize],
-    args: &[Vec<u8>],
-    func: &LoadedFunctionInstantiation,
-) -> Result<(), VMStatus> {
-    for (idx, (ty, arg)) in func.parameters.iter().zip(args.iter()).enumerate() {
-        if !idxs.contains(&idx) {
-            continue;
-        }
-        let arg_len = arg.len();
-        let mut cursor = Cursor::new(&arg[..]);
-        validate_arg(session, ty, &mut cursor, arg_len)?;
+fn validate_and_construct<S: MoveResolverExt>(
```

Comment:
Can you add some more documentation on cursor and how the recursion here works as there are multiple places where recursion happens:
1. In recursively_construct_arg where recursively_construct_arg is called
2. validate_and_construct calls recursively_construct_arg when there are multiple type args which can call back into recursively_construct_arg

Reply:
I added comments on how the cursor, re-serialization works